### PR TITLE
(PUP-10296) Add version range support to pip package provider

### PR DESCRIPTION
**Pull request by luchihoratiu**
Add support of `>`, `>=`, `<`,` <=`, `>=A <=B` ranges for package version specified in the :ensure manifest field. This will install a version that satisfies the provided range. The package version must respect pip's package versioning format (https://www.python.org/dev/peps/pep-0440/).

### CODE REVIEW

File: lib/puppet/provider/package/pip.rb [hunk 1]
```diff
@@ -116,99 +118,148 @@ def query
...
+          end
+        rescue VersionRange::ValidationFailure, PipVersion::ValidationFailure
+          Puppet.debug("Cannot parse #{should} as a pip version range, falling through")
+          
+          if should.is_a? String
+            command_options << "#{@resource[:name]}==#{should}"
+          else
+            command_options << @resource[:name]
+          end
+        end
```
Comment (GabrielNagy) [inline lib/puppet/provider/package/pip.rb]: I believe an approach similar to the regex match here https://github.com/puppetlabs/puppet/pull/8030/files#diff-0df6fa18d947b3857abe3ff1c68dfd44R318 would be better

I think it's more readable to keep the switch and do something like:

```ruby
when :latest ... # (do latest stuff)
when /[<>=]/ ... # (do version range stuff)
when String ... # (do normal version stuff)
else ... # (do no version stuff)
```

File: lib/puppet/provider/package/pip.rb [hunk 2]
```diff
@@ -226,6 +277,28 @@ def install_options
     join_options(@resource[:install_options])
   end
 
+  def insync?(is)
+    return false unless is && is != :absent
+
+    should = @resource[:ensure]
+
+    begin
```
Comment (GabrielNagy) [inline lib/puppet/provider/package/pip.rb]: Similar to my comment above, I would preemptively check if `is` and `should` match `/<>=/` before doing version range stuff.

File: lib/puppet/provider/package/pip.rb (hunk 1 above)
Comment (ciprianbadescu) [reply lib/puppet/provider/package/pip.rb]: I don't think provider should know about how a range is defined.
It should try to parse `should` as a Range and rescue ValidationError

Comment (GabrielNagy) [reply lib/puppet/provider/package/pip.rb]: But then you would try to parse a normal version like `1.2.3` when you don't actually want to.

Comment (GabrielNagy) [reply lib/puppet/provider/package/pip.rb]: so, after chatting we concluded that we shouldn't do a `/[<>=]/` check, only an `is_a?(String)` check before attempting to parse as a version range.

Comment (GabrielNagy) [reply lib/puppet/provider/package/pip.rb]: the same thing applies to the parsing inside the `insync` method

**Comment by luchihoratiu** [comment]
jenkins please test this

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos6-32a,centos8-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos6-32a,centos8-64a

File: lib/puppet/util/package/version/pip.rb [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    def self.parse(version)
+      new(version)
+    end
+
+    def self.compare(version_a, version_b)
+      version_a = new(version_a) unless version_a.is_a?(self.class)
+      version_b = new(version_b) unless version_a.is_a?(self.class)
+
+      version_a <=> version_b
+    end
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: i don't think this class should be responsible to compare to version. Versions should just be obejects comparable one to another

File: lib/puppet/util/package/version/pip.rb [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      parts.push(".post#{post_release[1]}")     if post_release
+      parts.push(".dev#{dev_release[1]}")       if dev_release
+      parts.push("+#{local_version.join(".")}") if local_version
+
+      parts.join
+    end
+
+    def <=>(other)
+      raise ValidationFailure.new(other.to_s) unless other.is_a?(self.class)
+      compare(self.key, other.key)
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: `self` is redundant here

File: lib/puppet/provider/package/pip.rb [hunk 5]
```diff
@@ -116,99 +118,146 @@ def query
...
         return nil
       end
     end
   end
 
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PipVersion.new(version)
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb]: why using `new` instead of `parse`?

File: lib/puppet/util/package/version/pip.rb [hunk 6]
```diff
@@ -0,0 +1,230 @@
+module Puppet::Util::Package::Version
+  class Pip
+    def initialize(version)
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: we should be consistent between Version Classes, apt has the `initialize` method private and expose only parse.

File: lib/puppet/util/package/version/pip.rb [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    end
+
+    def is_post_release?
+      post_release != nil
+    end
+
+    def is_dev_release?
+      dev_release != nil
+    end
+
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: why are this needed? i don't see them used anywhere
Alos in ruby you is a convention that you don't prefix wiht `is|has` just use the pridicate `pre_release?`

File: lib/puppet/util/package/version/pip.rb [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    private
+
+    @key = nil
+
+    def key=(val)
+      @key = val
+    end
+
+    def epoch
+      @epoch_data if @epoch_data != 0 && @epoch_data
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: i would check first for `@epoch_data`, it makes more sense to check for existence before checking for specific value

File: lib/puppet/util/package/version/pip.rb [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+          else
+            if ["c", "pre", "preview"].include?(letter)
+              letter = "rc"
+            else
+              if ["rev", "r"].include?(letter)
+                letter = "post"
+              end
+            end
+          end
+        end
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: you should youse `else/elsif` here instead of nested `ifs`

File: lib/puppet/util/package/version/pip.rb [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+        return [letter, number.to_i]
+      end
+
+      if !letter && number
+        letter = "post"
+        return [letter, number.to_i]
+      end
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: if there is no else, is better suited to use a guard clause:
https://anthonygharvey.com/ruby/guard_clauses_vs_nested_if_statements

File: lib/puppet/util/package/version/pip.rb [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    end
+
+    def parse_local_version(local_version)
+      return local_version.split(/[\\._-]/).map{|part| part =~ /[0-9]+/ && part !~ /[a-zA-Z]+/ ? part.to_i : part.downcase} if local_version
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: redundant `return`

File: lib/puppet/util/package/version/pip.rb [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    end
+
+    def compose_key(epoch, release, pre, post, dev, local)
+      release_key = release.reverse
+      release_key.each_with_index do |element, index|
+        break unless release_key.at(index) == 0
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: space after break

File: lib/puppet/util/package/version/pip.rb [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        release_key.delete_at(index) unless release_key.at(index + 1) != 0
+      end
+      release_key.reverse!
+
+      if !pre && !post && dev
+        pre_key = -Float::INFINITY
+      else
+        pre_key = pre ? pre : Float::INFINITY
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: use double `||` instead of `?`
```ruby
pre_key = pre || Float::INFINITY
```

File: lib/puppet/util/package/version/pip.rb [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+      end
+
+      post_key = post ? post : -Float::INFINITY
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: use double `||` instead of `?`

File: lib/puppet/util/package/version/pip.rb [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+      dev_key = dev ? dev : Float::INFINITY
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: use double `||` instead of `?`

File: lib/puppet/util/package/version/pip.rb [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+
+      if !local
+        local_key = -Float::INFINITY
+      else
+        local_key = local.map{|i| (i.is_a? Integer) ? [i, ""] : [-Float::INFINITY, i]}
+      end
+
+      return [epoch, release_key, pre_key, post_key, dev_key, local_key]
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: redundant `return`

File: lib/puppet/util/package/version/pip.rb [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+    def compare(this, other)
+      if (this.is_a? Array) && (other.is_a? Array)
+        this  << -Float::INFINITY if this.length < other.length
+        other << -Float::INFINITY if this.length > other.length
+
+        this.each_with_index do |element, index|
+          return compare(element, other.at(index)) if element != other.at(index)
+        end
+      elsif (this.is_a? Array) && !(other.is_a? Array)
+        raise "Cannot compare #{this} (Array) with #{other} (#{other.class}). Only ±Float::INFINITY accepted." unless other.abs == Float::INFINITY
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: this will raise RuntimeError, i think would be better to raise a specific error for this.

File: lib/puppet/util/package/version/pip.rb [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+        return this.at(0).to_f <=> other
+      elsif !(this.is_a? Array) && (other.is_a? Array)
+        raise "Cannot compare #{this} (#{this.class}) with #{other} (Array). Only ±Float::INFINITY accepted." unless this.abs == Float::INFINITY
+        return this <=> other.at(0).to_f
+      else
+        return this <=> other
+      end
+      return 0
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: redundant `return`

File: lib/puppet/util/package/version/pip.rb [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    end
+
+    @@VERSION_PATTERN = "
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: this should be a constant

File: lib/puppet/util/package/version/pip.rb [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+      @post_data
+    end
+
+    def dev_release
+      @dev_data
+    end
+
+    def local_version
+      @local_data
+    end
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: why do you need methods to return instance variable? you can create private readers

File: lib/puppet/util/package/version/pip.rb [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+
+    def self.compare(version_a, version_b)
+      version_a = new(version_a) unless version_a.is_a?(self.class)
+      version_b = new(version_b) unless version_a.is_a?(self.class)
+
+      version_a <=> version_b
+    end
+
+    def to_s
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: AptVersion also defines an alias for `to_s` as inspect. Should  Pip also have it, for consistency

File: lib/puppet/util/package/version/pip.rb [hunk 22]
```diff
@@ -0,0 +1,230 @@
+module Puppet::Util::Package::Version
+  class Pip
+    def initialize(version)
+      matched = version.match(Regexp.new(("^\\s*") + @@VERSION_PATTERN + ("\\s*$"), Regexp::EXTENDED | Regexp::MULTILINE | Regexp::IGNORECASE))
+
+      raise ValidationFailure.new(version) unless matched
+      self.class.include Comparable
```
[the remark this answers is not in the document]
Comment (luchihoratiu) [reply lib/puppet/util/package/version/pip.rb]: I thought it would be good to include Comparable only if the object will be created(string was parsed successfully) but I don't really know if there are any noticeable performance improvements.

File: lib/puppet/util/package/version/pip.rb (hunk 3 above)
Comment (luchihoratiu) [reply lib/puppet/util/package/version/pip.rb]: After chatting about this, it will be kept for now.

File: lib/puppet/util/package/version/pip.rb (hunk 7 above)
Comment (luchihoratiu) [reply lib/puppet/util/package/version/pip.rb]: Removed.

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

File: lib/puppet/provider/package/pip.rb [hunk 23]
```diff
@@ -166,49 +188,79 @@ def latest_with_old_pip
...
         process.collect do |line|
           # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
+          if line =~ /Using version .+? \(newest of versions: (.+?)\)/
+            textAfterLastMatch = $1.chomp(")\n")
+            versionList = textAfterLastMatch.split(', ').sort do |x,y|
+              Puppet::Util::Package.versioncmp(x, y)
+            end
+            return versionList
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb]: Before, `latest_with_old_pip` was returning the version without sorting avaialable versions. Will this change the behaviour, now that we are sorting the versions

File: lib/puppet/util/package/version/pip.rb [hunk 24]
```diff
@@ -0,0 +1,170 @@
+module Puppet::Util::Package::Version
+  class Pip
+    include Comparable
+
+    def self.parse(version)
+      raise ValidationFailure.new(version.to_s) unless version.is_a? String
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: we should be consistent when raising errors, see line: 122
i suggest using `raise(ErrorClass, message)` for better readability.

File: lib/puppet/util/package/version/pip.rb [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+    end
+
+    def parse_local_version(local_version)
+      local_version.split(/[\\._-]/).map{|part| part =~ /[0-9]+/ && part !~ /[a-zA-Z]+/ ? part.to_i : part.downcase} if local_version
+    end
+
+    def compose_key(epoch, release, pre, post, dev, local)
+      release_key = release.reverse
+      release_key.each_with_index do |element, index|
+        break unless release_key.at(index) == 0
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: why not use `element` here? which is the value at current index

File: lib/puppet/util/package/version/pip.rb [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+
+      matched = version.match(Regexp.new(("^\\s*") + VERSION_PATTERN + ("\\s*$"), Regexp::EXTENDED | Regexp::MULTILINE | Regexp::IGNORECASE))
+      raise ValidationFailure.new(version) unless matched
+
+      new(matched)
+    end
+
+    def self.compare(version_a, version_b)
+      version_a = parse(version_a) unless version_a.is_a?(Puppet::Util::Package::Version::Pip)
+      version_b = parse(version_b) unless version_a.is_a?(Puppet::Util::Package::Version::Pip)
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: you can use `self` directly here: `version_a.is_a?(self)`

File: lib/puppet/util/package/version/pip.rb [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+      if (this.is_a? Array) && (other.is_a? Array)
+        this  << -Float::INFINITY if this.length < other.length
+        other << -Float::INFINITY if this.length > other.length
+
+        this.each_with_index do |element, index|
+          return compare(element, other.at(index)) if element != other.at(index)
+        end
+      elsif (this.is_a? Array) && !(other.is_a? Array)
+        raise Puppet::Error, ("Cannot compare #{this} (Array) with #{other} (#{other.class}). Only ±Float::INFINITY accepted.") unless other.abs == Float::INFINITY
+        return this.at(0).to_f <=> other
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: `this.first`

File: lib/puppet/util/package/version/pip.rb [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+      elsif !(this.is_a? Array) && (other.is_a? Array)
+        raise Puppet::Error, ("Cannot compare #{this} (#{this.class}) with #{other} (Array). Only ±Float::INFINITY accepted.") unless this.abs == Float::INFINITY
+        return this <=> other.at(0).to_f
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: `other.first`

File: spec/unit/util/package/version/pip_spec.rb [hunk 29]
```diff
@@ -0,0 +1,439 @@
...
+
+      it "#{input_version} should be comparable" do
+        version = described_class.parse(input_version)
+
+        # rubocop:disable UselessComparison
+        expect(version <  version).to eq(false)
+        expect(version >  version).to eq(false)
+        expect(version != version).to eq(false)
+        expect(version == version).to eq(true)
+      end
```
Comment (gimmyxd) [inline spec/unit/util/package/version/pip_spec.rb]: i would extract this in 4 'it' examples, for each operator

File: spec/unit/util/package/version/pip_spec.rb [hunk 30]
```diff
@@ -0,0 +1,439 @@
...
+
+      it "#{invalid_input} should not be comparable with versions" do
+        valid_version = described_class.parse("1.0")
+
+        expect{ valid_version <  invalid_input }.to raise_error(described_class::ValidationFailure)
+        expect{ valid_version >  invalid_input }.to raise_error(described_class::ValidationFailure)
+        expect{ valid_version >= invalid_input }.to raise_error(described_class::ValidationFailure)
+        expect{ valid_version <= invalid_input }.to raise_error(described_class::ValidationFailure)
+      end
```
Comment (gimmyxd) [inline spec/unit/util/package/version/pip_spec.rb]: i would extract this in 4 'it' examples, for each operator

File: lib/puppet/util/package/version/pip.rb [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+          [-_\\.]?
+          (?<dev_l>dev)
+          [-_\\.]?
+          (?<dev_n>[0-9]+)?
+        )?
+      )
+      (?:\\+(?<local>[a-z0-9]+(?:[-_\\.][a-z0-9]+)*))?      # local version
+    "
+
+    class ValidationFailure < StandardError
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb]: This should inherit form `ArgumentError` to be consistent with `debian version`

File: spec/unit/util/package/version/pip_spec.rb [hunk 32]
```diff
@@ -0,0 +1,439 @@
...
+        end
+
+        it "#{lower_version} should be lower than #{greater_version}" do
+          expect(lower_version < greater_version).to eq(true)
+        end
+
+        it "#{greater_version} should be greater than #{lower_version}" do
+          expect(greater_version > lower_version).to eq(true)
+        end
+      end
```
Comment (gimmyxd) [inline spec/unit/util/package/version/pip_spec.rb]: you can use Array#combination to make testing pairs.

https://apidock.com/ruby/Array/combination

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

File: lib/puppet/provider/package/pip.rb [hunk 33]
```diff
@@ -166,49 +188,79 @@ def latest_with_old_pip
...
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = VersionRange.parse(should, PipVersion)
+          should = best_version(should_range)
+
+          if should
+            command_options << "#{@resource[:name]}==#{should}"
+            @resource[:ensure] = should.to_s
+          end
```
[the remark this answers is not in the document]
Comment (gimmyxd) [reply lib/puppet/provider/package/pip.rb]: Then should't the `fail` from :269 be here on the else branch?
This seems the only place where should is reassigned.

File: lib/puppet/provider/package/pip.rb [hunk 34]
```diff
@@ -11,7 +13,10 @@
   This provider supports the `install_options` attribute, which allows command-line flags to be passed to pip.
   These options should be specified as an array where each element is either a string or a hash."
 
-  has_feature :installable, :uninstallable, :upgradeable, :versionable, :install_options, :targetable
+  has_feature :installable, :uninstallable, :upgradeable, :versionable, :version_ranges, :install_options, :targetable
+
+  PipVersion   = Puppet::Util::Package::Version::Pip
+  VersionRange = Puppet::Util::Package::Version::Range
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb]: this will conflict with apt.rb VersionRagne
```
puppet/lib/puppet/provider/package/apt.rb:9: warning: already initialized constant VersionRange
puppet/lib/puppet/provider/package/pip.rb:19: warning: previous definition of VersionRange was here
```

File: lib/puppet/provider/package/pip.rb (hunk 33 above)
[the remark this answers is not in the document]
Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb]: Moved as discussed.

File: lib/puppet/provider/package/pip.rb [hunk 35]
```diff
@@ -166,38 +202,67 @@ def latest_with_old_pip
...
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should
+            command_options << "#{@resource[:name]}==#{should}"
+            @resource[:ensure] = should.to_s
```
Comment (luchihoratiu) [inline lib/puppet/provider/package/pip.rb]: After discussing with @gimmyxd, we think it's worth mentioning the role of this line of code. When ensuring the installation of a python package using version ranges, the output would normally be something like:
> Notice: /Stage[main]/Main/Package[Flask]/ensure: ensure changed '0.12.5' to _**'>1'**_

But by adding this line of code, the same output becomes:
> Notice: /Stage[main]/Main/Package[Flask]/ensure: ensure changed '0.12.5' to _**'1.1.1'**_

The output is given by the `execute` method which uses said `@resource[:ensure]` variable.
@puppetlabs/night-s-watch: I personally found this more useful (from  user perspective) than seeing the version range given as you don't need to run additional commands, besides the puppet run, to find out what version was installed. There shouldn't be any impact on the existing functionality but feel free to address any concerns here.

File: lib/puppet/provider/package/pip.rb [hunk 36]
```diff
@@ -226,6 +292,28 @@ def install_options
     join_options(@resource[:install_options])
   end
 
+  def insync?(is)
+    return false unless is && is != :absent
+    return false unless String === is && String === @resource[:ensure]
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb]: Do we need this check? seems that the `:absent` case is already handled and `resouce[:ensure]` is always a String

File: lib/puppet/provider/package/pip.rb [hunk 37]
```diff
@@ -166,38 +202,69 @@ def latest_with_old_pip
...
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
+      case should
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should == should_range
```
[the remark this answers is not in the document]
Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb]: The scope here was for when no version that satisfies the given range is found. We agreed, for now at least, that we will let the provider handle this case and give the error output. 

We need to concatenate `pip install package` with either `==A` or the introduced operators such as `>A <B`. Without this if, we would end up with something like `==>A <B` which will throw a syntax error rather than telling the user that no package version in the given range was found.  

I opted to replace  `should == should_range` with `should.is_a?(PIP_VERSION_RANGE)`. I think this is more readable.

File: lib/puppet/provider/package/pip.rb [hunk 38]
```diff
@@ -166,38 +202,69 @@ def latest_with_old_pip
...
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PIP_VERSION.parse(version)
+      included_available_versions.push(version) if should_range.include?(version)
+    end
+
+    sorted_result = included_available_versions.sort { |x,y| x <=> y }
+    return sorted_result.to_a.last if sorted_result
```
[the remark this answers is not in the document]
Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb]: Was trying to cover a case I rediscovered while working on the gem provider. Should be fixed now.

File: lib/puppet/provider/package/pip.rb (hunk 35 above)
Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb]: After chatting a bit on this topic, it was decided to remove this line until community/UX feedback is received.

File: lib/puppet/provider/package/pip.rb [hunk 39]
```diff
@@ -166,38 +200,68 @@ def latest_with_old_pip
...
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
+      case should
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should.is_a?(PIP_VERSION_RANGE)
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb]: hmmm, this feels a bit weird, we are trying to find the best versions, and if we cannot do that we return a range, and then we pass that range to pip command?

But i don't think we get PIP_VERSION_RANGE


```
❯ bx puppet resource package 'colorize' ensure='>3' provider='pip3'


=> #<Puppet::Util::Package::Version::Range::Gt:0x00007f88c0523bb8 @version=3>
[6] pry(#<Puppet::Type::Package::ProviderPip3>)> should.class
=> Puppet::Util::Package::Version::Range::Gt
```

Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb]: After discussing the issue, we agreed on previous implementation with a comment explaining the reason. Also the MinMax scenario was discovered and added.

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

File: acceptance/tests/provider/package/pip.rb [hunk 40]
```diff
@@ -22,21 +26,42 @@
...
+      package_manifest1 = resource_manifest('package', package, { ensure: '<=1.1.0', provider: 'pip' } )
+      package_manifest2 = resource_manifest('package', package, { ensure: '<1.0.4',  provider: 'pip' } )
+
+      # Ensure the package is not already installed
+      on(agent, "#{pip_command} uninstall #{package} --disable-pip-version-check --yes", :accept_all_exit_codes => true)
+
+      # Make a fresh package install (with version lower than or equal to 1.1.0)
+      apply_manifest_on(agent, package_manifest1, :expect_changes => true) do
+        list = on(agent, "#{pip_command} list --disable-pip-version-check").stdout
+        assert_match(/#{package} \(/, list)
```
Comment (ciprianbadescu) [inline acceptance/tests/provider/package/pip.rb]: Here you could also assert in version installed

File: acceptance/tests/provider/package/pip.rb [hunk 41]
```diff
@@ -22,21 +26,42 @@
...
+      end
+
+      # Reapply same manifest and expect no changes
+      apply_manifest_on(agent, package_manifest1, :catch_changes => true)
+
+      # Reinstall over existing package (with version lower than 1.0.4) and expect changes
+      apply_manifest_on(agent, package_manifest2, :expect_changes => true) do
+        list = on(agent, "#{pip_command} list --disable-pip-version-check").stdout
+        assert_match(/#{package} \(/, list)
```
Comment (ciprianbadescu) [inline acceptance/tests/provider/package/pip.rb]: Save here, assert could check that expected version in installed

File: acceptance/tests/provider/package/pip.rb [hunk 42]
```diff
@@ -22,21 +26,42 @@
...
     end
 
-    step "Install a pip package" do
+    step "Ensure presence of a pip package" do
       package_manifest = resource_manifest('package', package, { ensure: 'present', provider: 'pip' } )
       apply_manifest_on(agent, package_manifest, :catch_failures => true) do
         list = on(agent, "#{pip_command} list --disable-pip-version-check").stdout
         assert_match(/#{package} \(/, list)
       end
-      on(agent, "#{pip_command} uninstall #{package} --disable-pip-version-check --yes")
```
Comment (ciprianbadescu) [inline acceptance/tests/provider/package/pip.rb]: Why have you changed this? I find the behavior that a step removes a package it installed fair enough

**Comment by ciprianbadescu** [comment]
I tested and was working properly. OK to merge after test comments are fixed

File: acceptance/tests/provider/package/pip.rb (hunk 42 above)
Comment (luchihoratiu) [reply acceptance/tests/provider/package/pip.rb]: I initially thought that cleaning up after doing all the steps, through the added `teardown`, would be enough and I saw that line useless for the current combination of steps. 
But now that you've pointed it out, I agree that it would be nice to leave a clean environment after every step. It could get frustrating for someone who will want to modify or add new steps when random failures would occur.
Added the uninstall back 😁.

File: acceptance/tests/provider/package/pip.rb (hunk 40 above)
Comment (luchihoratiu) [reply acceptance/tests/provider/package/pip.rb]: Asserted

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a
